File: perl6/integ.pl
#!/usr/bin/env perl6

sub MAIN($nb) {
    if $nb < 10 || $nb > 100 { say "Bad option"; exit;  }
    say "n=" ~ $nb ~ "   Rectangles\tTrapezes\tSimpson";
    say "F1:    " ~ integ(1, $nb);
    say "F2:    " ~ integ(2, $nb);
    say "F3:    " ~ integ(3, $nb);
    say "F4:    " ~ integ(4, $nb);
    say "F5:    " ~ integ(5, $nb);
    say "F6:    " ~ integ(6, $nb);
    say "F7:    " ~ integ(7, $nb);
    say "F8:    " ~ integ(8, $nb);
    say "F9:    " ~ integ(9, $nb);
    say "F10:   " ~ integ(10, $nb);
}

sub integ($f, $n) {
    my $a = 0;
    my $b = 1;
    my $h = ($b - $a) / $n;

    my $string =
	sprintf("%.1E", rectangles($f, $n, $a, $h))   ~ "\t\t" ~
	sprintf("%.1E", trapezes($f, $n, $a, $b, $h)) ~ "\t\t" ~
	sprintf("%.1E", simpson($f, $n, $a, $b, $h))  ~ "\t\t";
    return $string
}

sub rectangles($f, $n, $a, $h) {
    my $s = 0;
    my $token = 0;

    while $token < $n {
	$s += f($f, $a + $h *  $token);
	$token++;
    }
   return abs (($h * $s) - 1)
}

sub trapezes($f, $n, $a, $b, $h) {
    my $s = 0;
    my $token = 1;

    while $token < $n {
	$s += f($f, $a + ($token * $h));
	$token++;
    }
    return abs(((($b - $a) / (2 * $n)) * (f($f, $a) + f($f, $b) + (2 * $s))) - 1)
}

sub simpson($f, $n, $a, $b, $h) {
    my $s1 = 0;
    my $s2 = 0;
    my $token = 1;

    while $token < $n {
	$s1 += f($f, $a + ($token * $h));
	$token++;
    }
    $token = 0;
    while $token < $n {
	$s2 += f($f, $a + ($token * $h) + ($h / 2));
	$token++;
    }
    return abs (((($b - $a) / (6 * $n)) * (f($f, $a) + f($f, $b) + (2 * $s1 + 4 * $s2))) - 1)
}

sub f($i, $x) {
    if $i == 1  {return 1}
    if $i == 2  {return 2 * $x}
    if $i == 3  {return 3 * ($x ** 2)}
    if $i == 4  {return 4 * ($x ** 3)}
    if $i == 5  {return 5 * ($x ** 4)}
    if $i == 6  {return 6 * ($x ** 5)}
    if $i == 7  {return 1 / (2 * (sqrt(2) - 1) * sqrt($x + 1))}
    if $i == 8  {return 2 / (($x + 1) ** 2)}
    if $i == 9  {return (pi / 2) * sin((pi / 2) * $x)}
    if $i == 10 {return (e ** $x) / (e - 1)}
}

File: lua/integ.lua
#!/usr/bin/env lua
require("math")

function main(n)
   print ("n=" .. tostring(n) ..  "   Rectangles\tTrapezes\tSimpson")
   print ("F1:\t" ..integ(1, n))
   print ("F2:\t" ..integ(2, n))
   print ("F3:\t" ..integ(3, n))
   print ("F4:\t" ..integ(4, n))
   print ("F5:\t" ..integ(5, n))
   print ("F6:\t" ..integ(6, n))
   print ("F7:\t" ..integ(7, n))
   print ("F8:\t" ..integ(8, n))
   print ("F9:\t" ..integ(9, n))
   print ("F10:\t" ..integ(10, n))
end

function integ (f, n)
   a = 0.0
   b = 1.0
   h = (b - a) / n
   string =
      string.format("%.1E\t\t", rectangles(f, n, a, h)) ..
      string.format("%.1E\t\t", trapezes(f, n, a, b, h)) ..
      string.format("%.1E\t\t", simpson(f, n, a, b, h))
   return string
end

function rectangles(f, n, a, h)
   s = 0.0
   token = 0.0

   while token < n do
      s = s + fu(f, (a + h * token))
      token = token + 1.0
   end
   return math.abs((h * s) - 1.0)
end

function trapezes(f, n, a, b, h)
   s = 0.0
   token = 1.0

   while token < n do
      s = s + fu(f, a + (token * h))
      token = token + 1.0
   end
   return math.abs((((b - a) / (2.0 * n)) * (fu(f, a) + fu(f, b) + (2.0 * s))) - 1.0)
end

function simpson(f, n, a, b, h)
   s1 = 0.0
   s2 = 0.0
   token = 1.0

    while token < n do
       s1 = s1 + fu(f, a + (token * h))
       token = token + 1.0
    end
    token = 0.0
    while token < n do
       s2 = s2 + fu(f, a + (token * h) + (h / 2))
       token = token + 1.0
    end
    return math.abs((((b - a) / (6.0 * n)) * (fu(f, a) + fu(f, b) + (2.0 * s1 + 4.0 * s2))) - 1.0)
end

function fu(i, x)
   if i == 1 then
      return 1.0
   elseif i == 2 then
      return (2.0 * x)
   elseif i == 3  then
      return (3.0 * math.pow(x, 2.0))
   elseif i == 4  then
      return 4.0 * math.pow(x, 3.0)
   elseif i == 5 then
      return 5.0 * math.pow(x, 4.0)
   elseif i == 6 then
      return 6.0 * math.pow(x, 5.0)
   elseif i == 7 then
      return 1.0 / (2.0 * (math.sqrt(2.0) - 1.0) * math.sqrt(x + 1.0))
   elseif i == 8 then
      return 2.0 / math.pow((x + 1.0), 2.0)
   elseif i == 9 then
      return (math.pi / 2.0) * math.sin((math.pi / 2.0) * x)
   elseif i == 10 then
      return math.pow(math.exp(1), x) / (math.exp(1) - 1)
   else
      return 1.0
   end
end

if #arg < 1
then
   print "Bad option"
   return
elseif tonumber(arg[1]) == nil then
   print "Bad option"
   return
elseif tonumber(arg[1]) < 10 then
   print "Bad option"
   return
elseif tonumber(arg[1]) > 100 then
   print "Bad option"
   return
end
main(tonumber(arg[1]))
